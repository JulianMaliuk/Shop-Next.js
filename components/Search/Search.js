import React, { Component } from 'react'
import { Input } from 'semantic-ui-react'
import _ from 'lodash';

// const debounce = (func, wait) => {
//   let timeout;

//   return function executedFunction(...args) {
//     const later = () => {
//       clearTimeout(timeout);
//       func(...args);
//     };

//     clearTimeout(timeout);
//     timeout = setTimeout(later, wait);
//   };
// };

export default class Search extends Component {

  state = {
    value: this.props.searchQuery,
    isLoading: false
  };

  runSearch = _.debounce((value) => {
    this.props.setSearchQuery(value);
    this.setState({ isLoading: false })
  }, 400)

  handleChange = ({target: {value}}) => {
    this.setState({ value, isLoading: true })
    this.runSearch(value);
  }

  clearSearch = () => {
    this.setState({ value: '' })
    this.props.setSearchQuery('');
  }

  render() {
    const { value, isLoading } = this.state
    return (
      <Input 
        action={{
          icon: 'remove',
          onClick: () => this.clearSearch(),
        }}
        fluid
        iconPosition='left'
        placeholder='Пошук...' 
        loading={isLoading}
        value={value} 
        onChange={this.handleChange} />
    )
  }
}
