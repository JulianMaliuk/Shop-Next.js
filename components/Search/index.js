import { connect } from 'react-redux';

import Search from './Search';
import {setSearchQuery} from "../../redux/reducers/products";

const mapStateToProps = (state) => ({ searchQuery: state.products.searchQuery })
  
const mapDispatchToProps = { setSearchQuery };

export default connect(mapStateToProps, mapDispatchToProps)(Search);
