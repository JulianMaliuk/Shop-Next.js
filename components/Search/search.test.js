import React from "react";
import { act } from "react-dom/test-utils";
import Enzyme, { shallow, render, mount } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

import Search from "./Search";

Enzyme.configure({ adapter: new Adapter() });

jest.useFakeTimers("modern")

describe('search component', () => {

  it('should pass a changed value to the onChange handler', async () => {
    const value = '2';
    const onChange = jest.fn();
    const wrapper = shallow(
      <Search setSearchQuery={onChange} searchQuery='' />
    );

    expect(wrapper.debug()).toMatchSnapshot();

    await act(async () => {
      wrapper.simulate('change', { target: { value: '1' } });
      wrapper.simulate('change', { target: { value: '2' } });
      wrapper.simulate('change', { target: { value } });

      jest.advanceTimersByTime(400)
    })

    expect(onChange).toBeCalledWith(value);
    expect(onChange).toHaveBeenCalledTimes(1);
  });

  it('should clear input by click button', () => {
    const wrapper = mount(
      <Search setSearchQuery={() => {}} searchQuery='text' />
    );

    wrapper.find('button').simulate('click');
    expect(wrapper.state().value).toBe('');
  });
  
  // it("changes value when clicked", () => {
      
  //   const onSearchMock = jest.fn();
  
  //   act(() => {
  //     render(<Search setSearchQuery={onSearchMock} searchQuery='' />, container);
  //   });
  
  //   const input = container.querySelector('input');
    
  //   Simulate.change(input, { target: { value: 'some data' } })
  
  //   expect(onSearchMock).not.toBeCalled();
  
  //   jest.advanceTimersByTime(500);
  //   jest.runAllTimers();

  //   expect(onSearchMock).toHaveBeenCalledTimes(1);

  //   // expect(onSearchMock).toBeCalledWith('some data');
  //   // expect(onSearchMock).toHaveBeenCalledTimes(1);
  // //   expect(button.innerHTML).toBe("Turn off");
  
  // //   act(() => {
  // //     for (let i = 0; i < 5; i++) {
  // //       button.dispatchEvent(new MouseEvent("click", { bubbles: true }));
  // //     }
  // //   });
  
  // //   expect(onChange).toHaveBeenCalledTimes(6);
  // //   expect(button.innerHTML).toBe("Turn on");
  // });
});
