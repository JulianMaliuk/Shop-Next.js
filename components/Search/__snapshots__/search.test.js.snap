// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`search component should pass a changed value to the onChange handler 1`] = `"<Input action={{...}} fluid={true} iconPosition=\\"left\\" placeholder=\\"Пошук...\\" loading={false} value=\\"\\" onChange={[Function (anonymous)]} type=\\"text\\" />"`;
